process/publisherHandler.go

```diff
@@ -0,0 +1,193 @@
+	cancelFunc func()
+	closeChan  chan struct{}
+}
+
+// NewPublisherHandler creates a new publisher handler component
+func NewPublisherHandler(
+	handler HyperBlockPublisher,
+	outportBlocksPool BlocksPool,
+	dataAggregator DataAggregator,
+	retryDurationInMiliseconds uint64,
+	firstCommitableBlocks map[uint32]uint64,
+) (*publisherHandler, error) {
+	if check.IfNil(handler) {
+		return nil, ErrNilPublisher
+	}
+	if check.IfNil(outportBlocksPool) {
+		return nil, ErrNilBlocksPool
+	}
+	if check.IfNil(dataAggregator) {
+		return nil, ErrNilDataAggregator
+	}
+
+	ph := &publisherHandler{
+		handler:               handler,
+		outportBlocksPool:     outportBlocksPool,
+		dataAggregator:        dataAggregator,
+		retryDuration:         time.Duration(retryDurationInMiliseconds) * time.Millisecond,
+		firstCommitableBlocks: firstCommitableBlocks,
+		blocksChan:            make(chan []byte),
+		closeChan:             make(chan struct{}),
+	}
+
+	var ctx context.Context
+	ctx, ph.cancelFunc = context.WithCancel(context.Background())
+
+	go ph.startListener(ctx)
+
+	return ph, nil
+}
+
+func (ph *publisherHandler) startListener(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug("closing commonPublisher listener")
+			return
+		case headerHash := <-ph.blocksChan:
+			ph.handlePublishEvent(headerHash)
+		}
+	}
+}
+
+// PublishHyperBlock will push aggregated outport block data to the firehose writer
+func (ph *publisherHandler) PublishBlock(headerHash []byte) error {
+	select {
+	case ph.blocksChan <- headerHash:
+	case <-ph.closeChan:
+	}
+
+	return nil
+}
+
+func (ph *publisherHandler) handlePublishEvent(headerHash []byte) {
+	for {
+		err := ph.handlerHyperOutportBlock(headerHash)
+		if err == nil {
+			return
+		}
+
+		log.Error("failed to publish hyper block event", "error", err)
+		time.Sleep(ph.retryDuration)
+	}
+}
+
+func (ph *publisherHandler) handlerHyperOutportBlock(headerHash []byte) error {
+	metaOutportBlock, err := ph.outportBlocksPool.GetMetaBlock(headerHash)
+	if err != nil {
+		return err
+	}
+
+	err = checkMetaOutportBlockHeader(metaOutportBlock)
+	if err != nil {
+		return err
+	}
+
+	metaNonce := metaOutportBlock.BlockData.Header.GetNonce()
+	shardID := metaOutportBlock.GetShardID()
+
+	if metaNonce < ph.firstCommitableBlocks[shardID] {
+		// do not try to aggregate or publish hyper outport block
+
+		log.Trace("do not commit block", "currentRound", metaNonce, "firstCommitableRound", ph.firstCommitableBlocks)
+
+		return nil
+	}
+
+	hyperOutportBlock, err := ph.dataAggregator.ProcessHyperBlock(metaOutportBlock)
+	if err != nil {
+		return err
+	}
+
+	lastCheckpoint, err := ph.getLastRoundsData(hyperOutportBlock)
+	if err != nil {
+		return fmt.Errorf("failed to get last round data: %w", err)
+	}
+
+	err = ph.handler.PublishHyperBlock(hyperOutportBlock)
+	if err != nil {
+		return fmt.Errorf("failed to publish hyperblock: %w", err)
+	}
+
+	return ph.outportBlocksPool.UpdateMetaState(lastCheckpoint)
```

Comment:
what happens if this operation fails, but PublishHyperBlock already happened?

Reply:
it will trigger publish again, good catch :+1: moved update state below so that if there is an error on update state it will not trigger publish multiple times